Darktrace/threat-visualizer/ingest/parser.yml

```diff
@@ -0,0 +1,54 @@
+name: threat-visualizer
+pipeline:
+  - name: json_event
+    external:
+      name: json.parse-json
+  - name: parsed_date
+    external:
+      name: date.parse
+      properties:
+        input_field: "{{json_event.message.time}}"
+        output_field: datetime
+        format: timestamp
+  - name: parsed_date_mod
+    external:
+      name: date.parse
+      properties:
+        input_field: "{{json_event.message.model.then.modified}}"
+        output_field: datetime
+        format: "%Y-%m-%d%H:%M:%S"
+  - name: set_common_fields
+  - name: set_custom_fields
+
+stages:
+  set_common_fields:
+    actions:
+      - set:
+          "@timestamp": "{{parsed_date.datetime}}"
+          event.category: "network"
+          event.end: "{{parsed_date_mod.datetime}}"
+          event.kind: "event"
+          event.type: ["info"]
+      - set:
+          event.kind: "alert"
+        filter: "{{json_event.message.model.then.actions.alert}}"
+      - set:
+          host.id: "{{json_event.message.device.did}}"
+          host.hostname: "{{json_event.message.device.hostname}}"
+          observer.name: "Darktrace"
+          observer.product: "Threat visualizer"
+          host.ip: >
+            [
+              {% for ip in json_event.message.device.ips %}
+              {% if ip.get("ip") %}
+              {% if ip["ip"] | is_ipaddress %}
+              '{{ip["ip"]}}',
+              {% endif %}
+              {% endif %}
+              {% endfor %}
+            ]
+
+  set_custom_fields:
+    actions:
+      - set:
+          darktrace.threat_visualizer: "{{json_event.message}}"
```

Comment:
Sadly, to avoid side effects, I suggest specifying all custom fields :( :
```suggestion
                    darktrace.threat_visualizer.creationTime: "{{json_event.message.creationTime}}"
          darktrace.threat_visualizer.commentCount: "{{json_event.message.commentCount}}"
          darktrace.threat_visualizer.pbid: "{{json_event.message.pbid}}"
          darktrace.threat_visualizer.time: "{{json_event.message.time}}"
          darktrace.threat_visualizer.model.then.name: "{{json_event.message.model.then.name}}"
          darktrace.threat_visualizer.model.then.pid: "{{json_event.message.model.then.pid}}"
          darktrace.threat_visualizer.model.then.phid: "{{json_event.message.model.then.phid}}"
          darktrace.threat_visualizer.model.then.uuid: "{{json_event.message.model.then.uuid}}"
          darktrace.threat_visualizer.model.then.mitre.tactics: "{{json_event.message.model.then.mitre.tactics}}"
          darktrace.threat_visualizer.model.then.mitre.techniques: "{{json_event.message.model.then.mitre.techniques}}"
          darktrace.threat_visualizer.model.then.tags: "{{json_event.message.model.then.tags}}"
          darktrace.threat_visualizer.model.then.description: "{{json_event.message.model.then.description}}"
          darktrace.threat_visualizer.model.then.behaviour: "{{json_event.message.model.then.behaviour}}"
          darktrace.threat_visualizer.model.then.defeats: "{{json_event.message.model.then.defeats}}"
          darktrace.threat_visualizer.model.then.defeats.arguments.value: "{{json_event.message.model.then.defeats.arguments.value}}"
          darktrace.threat_visualizer.model.then.defeats.comparator: "{{json_event.message.model.then.defeats.comparator}}"
          darktrace.threat_visualizer.model.then.defeats.defeatID: "{{json_event.message.model.then.defeats.defeatID}}"
          darktrace.threat_visualizer.model.then.defeats.filtertype: "{{json_event.message.model.then.defeats.filtertype}}"
          darktrace.threat_visualizer.model.then.version: "{{json_event.message.model.then.version}}"
          darktrace.threat_visualizer.model.now.name: "{{json_event.message.model.now.name}}"
          darktrace.threat_visualizer.model.now.pid: "{{json_event.message.model.now.pid}}"
          darktrace.threat_visualizer.model.now.phid: "{{json_event.message.model.now.phid}}"
          darktrace.threat_visualizer.model.now.uuid: "{{json_event.message.model.now.uuid}}"
          darktrace.threat_visualizer.model.now.mitre.tactics: "{{json_event.message.model.now.mitre.tactics}}"
          darktrace.threat_visualizer.model.now.mitre.techniques: "{{json_event.message.model.now.mitre.techniques}}"
          darktrace.threat_visualizer.model.now.tags: "{{json_event.message.model.now.tags}}"
          darktrace.threat_visualizer.model.now.description: "{{json_event.message.model.now.description}}"
          darktrace.threat_visualizer.model.now.behaviour: "{{json_event.message.model.now.behaviour}}"
          darktrace.threat_visualizer.model.now.defeats: "{{json_event.message.model.now.defeats}}"
          darktrace.threat_visualizer.model.now.defeats.arguments.value: "{{json_event.message.model.now.defeats.arguments.value}}"
          darktrace.threat_visualizer.model.now.defeats.comparator: "{{json_event.message.model.now.defeats.comparator}}"
          darktrace.threat_visualizer.model.now.defeats.defeatID: "{{json_event.message.model.now.defeats.defeatID}}"
          darktrace.threat_visualizer.model.now.defeats.filtertype: "{{json_event.message.model.now.defeats.filtertype}}"
          darktrace.threat_visualizer.model.now.edited.userID: "{{json_event.message.model.now.edited.userID}}"
          darktrace.threat_visualizer.model.now.message: "{{json_event.message.model.now.message}}"
          darktrace.threat_visualizer.model.now.priority: "{{json_event.message.model.now.priority}}"
          darktrace.threat_visualizer.model.now.category: "{{json_event.message.model.now.category}}"
          darktrace.threat_visualizer.model.now.version: "{{json_event.message.model.now.version}}"
          darktrace.threat_visualizer.score: "{{json_event.message.score}}"
          darktrace.threat_visualizer.device.ip: "{{json_event.message.device.ip}}"
          darktrace.threat_visualizer.device.ips: "{{json_event.message.device.ips}}"
          darktrace.threat_visualizer.device.ips.ip: "{{json_event.message.device.ips.ip}}"
          darktrace.threat_visualizer.device.ips.timems: "{{json_event.message.device.ips.timems}}"
          darktrace.threat_visualizer.device.ips.time: "{{json_event.message.device.ips.time}}"
          darktrace.threat_visualizer.device.ips.sid: "{{json_event.message.device.ips.sid}}"
          darktrace.threat_visualizer.device.sid: "{{json_event.message.device.sid}}"
          darktrace.threat_visualizer.device.firstSeen: "{{json_event.message.device.firstSeen}}"
          darktrace.threat_visualizer.device.lastSeen: "{{json_event.message.device.lastSeen}}"
          darktrace.threat_visualizer.device.typename: "{{json_event.message.device.typename}}"
          darktrace.threat_visualizer.device.typelabel: "{{json_event.message.device.typelabel}}"

```